sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/DecorrelateInnerQuery.scala

```diff
@@ -212,14 +214,38 @@ object DecorrelateInnerQuery extends PredicateHelper {
   }
 
   /**
-   * Rewrite all [[DomainJoin]]s in the inner query to actual inner joins with the outer query.
+   * Rewrite all [[DomainJoin]]s in the inner query to actual joins with the outer query.
    */
   def rewriteDomainJoins(
       outerPlan: LogicalPlan,
       innerPlan: LogicalPlan,
       conditions: Seq[Expression]): LogicalPlan = innerPlan match {
-    case d @ DomainJoin(domainAttrs, child) =>
+    case d @ DomainJoin(domainAttrs, child, joinType, condition) =>
       val domainAttrMap = buildDomainAttrMap(conditions, domainAttrs)
+
+      val newChild = joinType match {
+        // Left outer domain joins are used to handle the COUNT bug.
+        case LeftOuter =>
+          // Replace the attributes in the domain join condition with the actual outer expressions
+          // and use the new join conditions to rewrite domain joins in its child. For example:
+          // DomainJoin [c'] LeftOuter (a = c') with domainAttrMap: { c' -> _1 }.
+          // Then the new conditions to use will be [(a = _1)].
+          val newConditions = condition.map(
+            _.transform { case a: Attribute => domainAttrMap.getOrElse(a, a)}
+          ).map(splitConjunctivePredicates).getOrElse(conditions)
```

Comment:
If `condition` is not None, we totally ignore `conditions`?

Reply:
`condition` should only be None for Inner join. For left outer join the condition should never be None (otherwise the node is not correlated and thus shouldn't be reached).

Reply:
That said, for left outer domain join, we don't need to propagate the `conditions` parameter of `rewriteDomainJoins` when recursively calling `rewriteDomainJoins`?

Reply:
and let's make the code more explicit
```
assert(condition.isDefne)
val newConditions = condition.get.trasform ...
```